File: pages/home_page.py
import time
from init_helpers import *

FIND_A_PATIENT_NAV_ELEMENT = ("role", "link", "Find a patient")
VACCINES_NAV_ELEMENT = ("role", "link", "Vaccines")
REPORTS_NAV_ELEMENT = ("role", "link", "Reports")
MANAGE_USERS_NAV_ELEMENT = ("role", "link", "Manage users")
RECORD_VACCINATIONS_NAV_ELEMENT = ("role", "link", "Record vaccinations")
NHS_LOGO_NAV_ELEMENT = ("role", "link", "NHS Logo Record a vaccination")
LOGOUT_NAV_ELEMENT = ("role", "link", "Log out")
NAV_BAR_TOGGLER = ("//button[@class='navbar-toggler']")
NAV_LINK_BAR_TOGGLER = ("//button[@class='navbar-toggler p-2']")
FEEDBACK_LINK = ("role", "link", "feedback (opens in a new tab)")
REPORT_AN_ISSUE_LINK = ("role", "link", "Report an issue")
ADD_VACCINES_LINK = ("role", "link", "Add vaccines")
ADD_USERS_LINK = ("role", "link", "Add users")
FIND_A_PATIENT_LINK = ("nested_role", "link", "Find a patient", True, "#maincontent")
CONTACT_US_LINK = ("role", "link", "Contact us")
HELP_AND_GUIDANCE_LINK = ("role", "link", "Help and guidance")
USER_GUIDE_LINK = ("role", "link", "User guide (opens in new tab)")
TERMS_OF_USE_LINK = ("role", "link", "Terms of use")
COOKIES_LINK = ("role", "link", "Cookies")
ACCESSIBILITY_STATEMENT_LINK = ("role", "link", "Accessibility statement")
ADD_USER_BUTTON = ("role", "button", "Add user")
SEARCH_BUTTON = ("role", "button", "Search")
ADD_VACCINE_BUTTON = ("role", "button", "Add vaccine")
CREATE_REPORT_BUTTON = ("role", "button", "Create report")
PAGE_LOADING_ELEMENT = ("role", "status")
TODAY_VACCINATION_COUNT = ("id", "today-vaccinations")
WEEK_VACCINATION_COUNT = ("id", "week-vaccinations")
MONTH_VACCINATION_COUNT = ("id", "month-vaccinations")
CREATE_A_REPORT_LINK = ("role", "link", "create a report")
TOTAL_VACCINATIONS_TEXT_ELEMENT = ("role", "heading", "Total vaccinations")

def ensure_log_out_nav_element_exists():
    wait_for_element_to_disappear(PAGE_LOADING_ELEMENT)
    if not check_element_exists(LOGOUT_NAV_ELEMENT):
        wait_for_element_to_disappear(PAGE_LOADING_ELEMENT)
        wait_for_element_to_appear(LOGOUT_NAV_ELEMENT)

def check_feedback_link_exists():
    ensure_log_out_nav_element_exists()
    return check_element_exists(FEEDBACK_LINK)

def check_accessibility_statement_link_exists():
    ensure_log_out_nav_element_exists()
    return check_element_exists(ACCESSIBILITY_STATEMENT_LINK)

def check_terms_of_use_link_exists():
    ensure_log_out_nav_element_exists()
    return check_element_exists(TERMS_OF_USE_LINK)

def check_user_guide_link_exists():
    ensure_log_out_nav_element_exists()
    return check_element_exists(USER_GUIDE_LINK)

def click_accessibility_statement_link():
    ensure_log_out_nav_element_exists()
    find_element_and_perform_action(ACCESSIBILITY_STATEMENT_LINK, "click")

def click_user_guide_link():
    ensure_log_out_nav_element_exists()
    find_element_and_perform_action(USER_GUIDE_LINK, "click")

def click_terms_of_use_link():
    ensure_log_out_nav_element_exists()
    find_element_and_perform_action(TERMS_OF_USE_LINK, "click")

def check_add_vaccines_link_exists():
    ensure_log_out_nav_element_exists()
    return check_element_exists(ADD_VACCINES_LINK)

def check_add_users_link_exists():
    ensure_log_out_nav_element_exists()
    return check_element_exists(ADD_USERS_LINK)

def check_find_a_patient_link_exists():
    ensure_log_out_nav_element_exists()
    return check_element_exists(FIND_A_PATIENT_LINK)

def click_add_vaccines_link():
    ensure_log_out_nav_element_exists()
    find_element_and_perform_action(ADD_VACCINES_LINK, "click")

def click_add_users_link():
    ensure_log_out_nav_element_exists()
    find_element_and_perform_action(ADD_USERS_LINK, "click")

def click_find_a_patient_link():
    ensure_log_out_nav_element_exists()
    find_element_and_perform_action(FIND_A_PATIENT_LINK, "click")

def check_site_name_exists_in_dashboard(site):
    ensure_log_out_nav_element_exists()
    element = ("role", "heading", site)
    return check_element_exists(element)

def get_today_vaccinations_count():
    ensure_log_out_nav_element_exists()
    text = find_element_and_perform_action(TODAY_VACCINATION_COUNT, "get_text")
    count = re.findall(r'\d+', text)
    return count[0]

def get_week_vaccinations_count():
    ensure_log_out_nav_element_exists()
    text = find_element_and_perform_action(WEEK_VACCINATION_COUNT, "get_text")
    count = re.findall(r'\d+', text)
    return count[0]

def get_month_vaccinations_count():
    ensure_log_out_nav_element_exists()
    text = find_element_and_perform_action(MONTH_VACCINATION_COUNT, "get_text")
    count = re.findall(r'\d+', text)
    return count[0]

def check_create_a_report_link_exists():
    ensure_log_out_nav_element_exists()
    return check_element_exists(CREATE_A_REPORT_LINK)

def click_feedback_link_exists():
    ensure_log_out_nav_element_exists()
    wait_for_element_to_appear(FEEDBACK_LINK)
    find_element_and_perform_action(FEEDBACK_LINK, "click")
    wait_for_element_to_disappear(PAGE_LOADING_ELEMENT)

def check_report_an_issue_link_exists():
    ensure_log_out_nav_element_exists()
    wait_for_element_to_appear(REPORT_AN_ISSUE_LINK)
    return check_element_exists(REPORT_AN_ISSUE_LINK)

def click_report_an_issue_link_exists():
    ensure_log_out_nav_element_exists()
    wait_for_element_to_appear(REPORT_AN_ISSUE_LINK)
    find_element_and_perform_action(REPORT_AN_ISSUE_LINK, "click")
    wait_for_element_to_disappear(PAGE_LOADING_ELEMENT)

def check_contact_us_link_exists():
    ensure_log_out_nav_element_exists()
    wait_for_element_to_appear(CONTACT_US_LINK)
    return check_element_exists(CONTACT_US_LINK)

def click_contact_us_link_exists():
    ensure_log_out_nav_element_exists()
    wait_for_element_to_appear(CONTACT_US_LINK)
    find_element_and_perform_action(CONTACT_US_LINK, "click")
    wait_for_element_to_disappear(PAGE_LOADING_ELEMENT)

def check_help_and_guidance_link_exists():
    ensure_log_out_nav_element_exists()
    wait_for_element_to_appear(HELP_AND_GUIDANCE_LINK)
    return check_element_exists(HELP_AND_GUIDANCE_LINK)

def click_help_and_guidance_link_exists():
    ensure_log_out_nav_element_exists()
    wait_for_element_to_appear(HELP_AND_GUIDANCE_LINK)
    find_element_and_perform_action(HELP_AND_GUIDANCE_LINK, "click")

def click_logout_button():
    ensure_log_out_nav_element_exists()
    find_element_and_perform_action(LOGOUT_NAV_ELEMENT, "click")
    wait_for_element_to_disappear(PAGE_LOADING_ELEMENT)
    time.sleep(5)

def click_navbar_toggler():
    ensure_log_out_nav_element_exists()
    wait_for_element_to_appear(NAV_BAR_TOGGLER)
    find_element_and_perform_action(NAV_BAR_TOGGLER, "click")

def click_nav_link_bar_toggler():
    ensure_log_out_nav_element_exists()
    wait_for_element_to_appear(NAV_LINK_BAR_TOGGLER)
    find_element_and_perform_action(NAV_LINK_BAR_TOGGLER, "click")

def check_logout_button_exists():
    ensure_log_out_nav_element_exists()
    return check_element_exists(LOGOUT_NAV_ELEMENT, True)

def check_navbar_toggle_exists_without_waiting():
    ensure_log_out_nav_element_exists()
    return check_element_exists(NAV_BAR_TOGGLER, False)

def check_navbar_toggle_exists():
    ensure_log_out_nav_element_exists()
    wait_for_element_to_appear(NAV_BAR_TOGGLER)
    return check_element_exists(NAV_BAR_TOGGLER, True)

def check_nav_link_bar_toggle_exists_without_waiting():
    ensure_log_out_nav_element_exists()
    return check_element_exists(NAV_LINK_BAR_TOGGLER, False)

def check_nav_link_bar_toggle_exists():
    ensure_log_out_nav_element_exists()
    wait_for_element_to_appear(NAV_LINK_BAR_TOGGLER)
    return check_element_exists(NAV_LINK_BAR_TOGGLER, True)

def check_logout_button_exists_without_waiting():
    ensure_log_out_nav_element_exists()
    return check_element_exists(LOGOUT_NAV_ELEMENT, False)

def click_profile_nav_link(email):
    # ensure_log_out_nav_element_exists()
    ensure_log_out_nav_element_exists()
    element = ("role", "link", email)
    wait_for_element_to_appear(element)
    find_element_and_perform_action(element, "click")
    wait_for_element_to_disappear(PAGE_LOADING_ELEMENT)

def check_reports_nav_link_exists():
    ensure_log_out_nav_element_exists()
    wait_for_element_to_appear(FIND_A_PATIENT_NAV_ELEMENT)
    return check_element_exists(REPORTS_NAV_ELEMENT)

def click_reports_nav_link():
    ensure_log_out_nav_element_exists()
    wait_for_element_to_appear(REPORTS_NAV_ELEMENT)
    find_element_and_perform_action(REPORTS_NAV_ELEMENT, "click")
    wait_for_element_to_disappear(PAGE_LOADING_ELEMENT)
    wait_for_element_to_appear(CREATE_REPORT_BUTTON)

def check_vaccines_nav_link_exists():
    ensure_log_out_nav_element_exists()
    wait_for_element_to_appear(FIND_A_PATIENT_NAV_ELEMENT)
    return check_element_exists(VACCINES_NAV_ELEMENT)

def click_vaccines_nav_link():
    ensure_log_out_nav_element_exists()
    wait_for_element_to_appear(VACCINES_NAV_ELEMENT)
    find_element_and_perform_action(VACCINES_NAV_ELEMENT, "click")
    wait_for_element_to_disappear(PAGE_LOADING_ELEMENT)
    wait_for_element_to_appear(ADD_VACCINE_BUTTON)

def check_manage_users_nav_link_exists():
    ensure_log_out_nav_element_exists()
    wait_for_element_to_appear(FIND_A_PATIENT_NAV_ELEMENT)
    return check_element_exists(MANAGE_USERS_NAV_ELEMENT)

def click_manage_users_nav_link():
    ensure_log_out_nav_element_exists()
    wait_for_element_to_appear(MANAGE_USERS_NAV_ELEMENT)
    find_element_and_perform_action(MANAGE_USERS_NAV_ELEMENT, "click")
    wait_for_element_to_disappear(PAGE_LOADING_ELEMENT)
    wait_for_element_to_appear(ADD_USER_BUTTON)

def check_record_vaccinations_nav_link_exists():
    ensure_log_out_nav_element_exists()
    wait_for_element_to_appear(FIND_A_PATIENT_NAV_ELEMENT)
    return check_element_exists(RECORD_VACCINATIONS_NAV_ELEMENT)

def click_record_vaccinations_nav_link():
    ensure_log_out_nav_element_exists()
    wait_for_element_to_appear(FIND_A_PATIENT_NAV_ELEMENT)
    find_element_and_perform_action(RECORD_VACCINATIONS_NAV_ELEMENT, "click")

def click_find_a_patient_nav_link():
    ensure_log_out_nav_element_exists()
    wait_for_element_to_appear(FIND_A_PATIENT_NAV_ELEMENT)
    find_element_and_perform_action(FIND_A_PATIENT_NAV_ELEMENT, "click")
    wait_for_element_to_disappear(PAGE_LOADING_ELEMENT)
    wait_for_element_to_appear(SEARCH_BUTTON)

File: ts-tests/e2e/pages/home_page.ts
import { playwrightHelperInstance } from '../init_helpers';

export const FIND_A_PATIENT_NAV_ELEMENT = { type: 'role', value: 'link', name: 'Find a patient' };
export const VACCINES_NAV_ELEMENT = { type: 'role', value: 'link', name: 'Vaccines' };
export const REPORTS_NAV_ELEMENT = { type: 'role', value: 'link', name: 'Reports' };
export const MANAGE_USERS_NAV_ELEMENT = { type: 'role', value: 'link', name: 'Manage users' };
export const RECORD_VACCINATIONS_NAV_ELEMENT = { type: 'role', value: 'link', name: 'Record vaccinations' };
export const NHS_LOGO_NAV_ELEMENT = { type: 'role', value: 'link', name: 'NHS Logo Record a vaccination' };
export const LOGOUT_NAV_ELEMENT = { type: 'role', value: 'link', name: 'Log out' };
export const NAV_BAR_TOGGLER = { type: 'xpath', value: "//button[@class='navbar-toggler']" };
export const NAV_LINK_BAR_TOGGLER = { type: 'xpath', value: "//button[@class='navbar-toggler p-2']" };
export const FEEDBACK_LINK = { type: 'role', value: 'link', name: 'feedback (opens in a new tab)' };
export const REPORT_AN_ISSUE_LINK = { type: 'role', value: 'link', name: 'Report an issue' };
export const ADD_VACCINES_LINK = { type: 'role', value: 'link', name: 'Add vaccines' };
export const ADD_USERS_LINK = { type: 'role', value: 'link', name: 'Add users' };
export const FIND_A_PATIENT_LINK = { type: 'nested_role', value: 'link', name: 'Find a patient', exact: true, parentSelector: '#maincontent' };
export const CONTACT_US_LINK = { type: 'role', value: 'link', name: 'Contact us' };
export const HELP_AND_GUIDANCE_LINK = { type: 'role', value: 'link', name: 'Help and guidance' };
export const USER_GUIDE_LINK = { type: 'role', value: 'link', name: 'User guide (opens in new tab)' };
export const TERMS_OF_USE_LINK = { type: 'role', value: 'link', name: 'Terms of use' };
export const COOKIES_LINK = { type: 'role', value: 'link', name: 'Cookies' };
export const ACCESSIBILITY_STATEMENT_LINK = { type: 'role', value: 'link', name: 'Accessibility statement' };
export const ADD_USER_BUTTON = { type: 'role', value: 'button', name: 'Add user' };
export const SEARCH_BUTTON = { type: 'role', value: 'button', name: 'Search' };
export const ADD_VACCINE_BUTTON = { type: 'role', value: 'button', name: 'Add vaccine' };
export const CREATE_REPORT_BUTTON = { type: 'role', value: 'button', name: 'Create report' };
export const PAGE_LOADING_ELEMENT = { type: 'role', value: 'status' };
export const TODAY_VACCINATION_COUNT = { type: 'id', value: 'today-vaccinations' };
export const WEEK_VACCINATION_COUNT = { type: 'id', value: 'week-vaccinations' };
export const MONTH_VACCINATION_COUNT = { type: 'id', value: 'month-vaccinations' };
export const CREATE_A_REPORT_LINK = { type: 'role', value: 'link', name: 'create a report' };
export const TOTAL_VACCINATIONS_TEXT_ELEMENT = { type: 'role', value: 'heading', name: 'Total vaccinations' };

export async function ensureLogOutNavElementExists(): Promise<void> {
    if (!(await playwrightHelperInstance!.checkElementExists(LOGOUT_NAV_ELEMENT))) {
        await playwrightHelperInstance!.waitForElementToDisappear(PAGE_LOADING_ELEMENT);
        await playwrightHelperInstance!.waitForElementToAppear(LOGOUT_NAV_ELEMENT);
    }
}

export async function checkFeedbackLinkExists(): Promise<boolean> {
    await ensureLogOutNavElementExists();
    return await playwrightHelperInstance!.checkElementExists({ type: 'role', value: 'link', name: 'feedback (opens in a new tab)' });
}

export async function checkAccessibilityStatementLinkExists(): Promise<boolean> {
    await ensureLogOutNavElementExists();
    return await playwrightHelperInstance!.checkElementExists({ type: 'role', value: 'link', name: 'Accessibility statement' });
}

export async function checkTermsOfUseLinkExists(): Promise<boolean> {
    await ensureLogOutNavElementExists();
    return await playwrightHelperInstance!.checkElementExists({ type: 'role', value: 'link', name: 'Terms of use' });
}

export async function checkUserGuideLinkExists(): Promise<boolean> {
    await ensureLogOutNavElementExists();
    return await playwrightHelperInstance!.checkElementExists({ type: 'role', value: 'link', name: 'User guide (opens in new tab)' });
}

export async function clickAccessibilityStatementLink(): Promise<void> {
    await ensureLogOutNavElementExists();
    await playwrightHelperInstance!.findElementAndPerformAction({ type: 'role', value: 'link', name: 'Accessibility statement' }, 'click');
}

export async function clickUserGuideLink(): Promise<void> {
    await ensureLogOutNavElementExists();
    await playwrightHelperInstance!.findElementAndPerformAction({ type: 'role', value: 'link', name: 'User guide (opens in new tab)' }, 'click');
}

export async function clickTermsOfUseLink(): Promise<void> {
    await ensureLogOutNavElementExists();
    await playwrightHelperInstance!.findElementAndPerformAction({ type: 'role', value: 'link', name: 'Terms of use' }, 'click');
}

export async function checkAddVaccinesLinkExists(): Promise<boolean> {
    await ensureLogOutNavElementExists();
    return await playwrightHelperInstance!.checkElementExists({ type: 'role', value: 'link', name: 'Add vaccines' });
}

export async function checkAddUsersLinkExists(): Promise<boolean> {
    await ensureLogOutNavElementExists();
    return await playwrightHelperInstance!.checkElementExists({ type: 'role', value: 'link', name: 'Add users' });
}

export async function checkFindAPatientLinkExists(): Promise<boolean> {
    await ensureLogOutNavElementExists();
    return await playwrightHelperInstance!.checkElementExists({ type: 'nested_role', value: 'link', name: 'Find a patient', parent: '#maincontent' });
}

export async function clickAddVaccinesLink(): Promise<void> {
    await ensureLogOutNavElementExists();
    await playwrightHelperInstance!.findElementAndPerformAction({ type: 'role', value: 'link', name: 'Add vaccines' }, 'click');
}

export async function clickAddUsersLink(): Promise<void> {
    await ensureLogOutNavElementExists();
    await playwrightHelperInstance!.findElementAndPerformAction({ type: 'role', value: 'link', name: 'Add users' }, 'click');
}

export async function clickFindAPatientLink(): Promise<void> {
    await ensureLogOutNavElementExists();
    await playwrightHelperInstance!.findElementAndPerformAction({ type: 'nested_role', value: 'link', name: 'Find a patient', parent: '#maincontent' }, 'click');
}

export async function checkSiteNameExistsInDashboard(site: string): Promise<boolean> {
    await ensureLogOutNavElementExists();
    const element = { type: 'role', value: 'heading', name: site };
    return await playwrightHelperInstance!.checkElementExists(element);
}

export async function getTodayVaccinationsCount(): Promise<string> {
    await ensureLogOutNavElementExists();
    const text = await playwrightHelperInstance!.findElementAndPerformAction(TODAY_VACCINATION_COUNT, 'get_text');
    const count = text.match(/\d+/);
    return count ? count[0] : '0';
}

export async function getWeekVaccinationsCount(): Promise<string> {
    await ensureLogOutNavElementExists();
    const text = await playwrightHelperInstance!.findElementAndPerformAction(WEEK_VACCINATION_COUNT, 'get_text');
    const count = text.match(/\d+/);
    return count ? count[0] : '0';
}

export async function getMonthVaccinationsCount(): Promise<string> {
    await ensureLogOutNavElementExists();
    const text = await playwrightHelperInstance!.findElementAndPerformAction(MONTH_VACCINATION_COUNT, 'get_text');
    const count = text.match(/\d+/);
    return count ? count[0] : '0';
}

export async function checkCreateAReportLinkExists(): Promise<boolean> {
    await ensureLogOutNavElementExists();
    return await playwrightHelperInstance!.checkElementExists(CREATE_A_REPORT_LINK);
}

export async function clickFeedbackLinkExists(): Promise<void> {
    await ensureLogOutNavElementExists();
    await playwrightHelperInstance!.waitForElementToAppear({ type: 'role', value: 'link', name: 'feedback (opens in a new tab)' });
    await playwrightHelperInstance!.findElementAndPerformAction({ type: 'role', value: 'link', name: 'feedback (opens in a new tab)' }, 'click');
    await playwrightHelperInstance!.waitForElementToDisappear(PAGE_LOADING_ELEMENT);
}

export async function checkReportAnIssueLinkExists(): Promise<boolean> {
    await ensureLogOutNavElementExists();
    await playwrightHelperInstance!.waitForElementToAppear({ type: 'role', value: 'link', name: 'Report an issue' });
    return await playwrightHelperInstance!.checkElementExists({ type: 'role', value: 'link', name: 'Report an issue' });
}

export async function clickReportAnIssueLinkExists(): Promise<void> {
    await ensureLogOutNavElementExists();
    await playwrightHelperInstance!.waitForElementToAppear({ type: 'role', value: 'link', name: 'Report an issue' });
    await playwrightHelperInstance!.findElementAndPerformAction({ type: 'role', value: 'link', name: 'Report an issue' }, 'click');
    await playwrightHelperInstance!.waitForElementToDisappear(PAGE_LOADING_ELEMENT);
}

export async function checkContactUsLinkExists(): Promise<boolean> {
    await ensureLogOutNavElementExists();
    await playwrightHelperInstance!.waitForElementToAppear({ type: 'role', value: 'link', name: 'Contact us' });
    return await playwrightHelperInstance!.checkElementExists({ type: 'role', value: 'link', name: 'Contact us' });
}

export async function clickContactUsLinkExists(): Promise<void> {
    await ensureLogOutNavElementExists();
    await playwrightHelperInstance!.waitForElementToAppear({ type: 'role', value: 'link', name: 'Contact us' });
    await playwrightHelperInstance!.findElementAndPerformAction({ type: 'role', value: 'link', name: 'Contact us' }, 'click');
    await playwrightHelperInstance!.waitForElementToDisappear(PAGE_LOADING_ELEMENT);
}

export async function checkHelpAndGuidanceLinkExists(): Promise<boolean> {
    await ensureLogOutNavElementExists();
    await playwrightHelperInstance!.waitForElementToAppear({ type: 'role', value: 'link', name: 'Help and guidance' });
    return await playwrightHelperInstance!.checkElementExists({ type: 'role', value: 'link', name: 'Help and guidance' });
}

export async function clickHelpAndGuidanceLinkExists(): Promise<void> {
    await ensureLogOutNavElementExists();
    await playwrightHelperInstance!.waitForElementToAppear({ type: 'role', value: 'link', name: 'Help and guidance' });
    await playwrightHelperInstance!.findElementAndPerformAction({ type: 'role', value: 'link', name: 'Help and guidance' }, 'click');
}

export async function clickLogoutButton(): Promise<void> {
    await ensureLogOutNavElementExists();
    await playwrightHelperInstance!.findElementAndPerformAction(LOGOUT_NAV_ELEMENT, 'click');
    await playwrightHelperInstance!.waitForElementToDisappear(PAGE_LOADING_ELEMENT);
}

export async function clickNavbarToggler(): Promise<void> {
    await ensureLogOutNavElementExists();
    await playwrightHelperInstance!.waitForElementToAppear({ type: 'xpath', value: "//button[@class='navbar-toggler']" });
    await playwrightHelperInstance!.findElementAndPerformAction({ type: 'xpath', value: "//button[@class='navbar-toggler']" }, 'click');
}

export async function clickNavLinkBarToggler(): Promise<void> {
    await ensureLogOutNavElementExists();
    await playwrightHelperInstance!.waitForElementToAppear({ type: 'xpath', value: "//button[@class='navbar-toggler p-2']" });
    await playwrightHelperInstance!.findElementAndPerformAction({ type: 'xpath', value: "//button[@class='navbar-toggler p-2']" }, 'click');
}

export async function checkLogoutButtonExists(): Promise<boolean> {
    await ensureLogOutNavElementExists();
    return await playwrightHelperInstance!.checkElementExists(LOGOUT_NAV_ELEMENT);
}

export async function checkNavbarToggleExistsWithoutWaiting(): Promise<boolean> {
    await ensureLogOutNavElementExists();
    return await playwrightHelperInstance!.checkElementExists({ type: 'xpath', value: "//button[@class='navbar-toggler']" });
}

export async function checkNavbarToggleExists(): Promise<boolean> {
    await ensureLogOutNavElementExists();
    await playwrightHelperInstance!.waitForElementToAppear({ type: 'xpath', value: "//button[@class='navbar-toggler']" });
    return await playwrightHelperInstance!.checkElementExists({ type: 'xpath', value: "//button[@class='navbar-toggler']" });
}

export async function checkNavLinkBarToggleExistsWithoutWaiting(): Promise<boolean> {
    await ensureLogOutNavElementExists();
    return await playwrightHelperInstance!.checkElementExists({ type: 'xpath', value: "//button[@class='navbar-toggler p-2']" });
}

export async function checkNavLinkBarToggleExists(): Promise<boolean> {
    await ensureLogOutNavElementExists();
    await playwrightHelperInstance!.waitForElementToAppear({ type: 'xpath', value: "//button[@class='navbar-toggler p-2']" });
    return await playwrightHelperInstance!.checkElementExists({ type: 'xpath', value: "//button[@class='navbar-toggler p-2']" });
}

export async function checkLogoutButtonExistsWithoutWaiting(): Promise<boolean> {
    await ensureLogOutNavElementExists();
    return await playwrightHelperInstance!.checkElementExists(LOGOUT_NAV_ELEMENT);
}

export async function clickProfileNavLink(email: string): Promise<void> {
    await ensureLogOutNavElementExists();
    const element = { type: 'role', value: 'link', name: email };
    await playwrightHelperInstance!.waitForElementToAppear(element);
    await playwrightHelperInstance!.findElementAndPerformAction(element, 'click');
    await playwrightHelperInstance!.waitForElementToDisappear(PAGE_LOADING_ELEMENT);
}

export async function checkReportsNavLinkExists(): Promise<boolean> {
    await ensureLogOutNavElementExists();
    await playwrightHelperInstance!.waitForElementToAppear(FIND_A_PATIENT_NAV_ELEMENT);
    return await playwrightHelperInstance!.checkElementExists(REPORTS_NAV_ELEMENT);
}

export async function clickReportsNavLink(): Promise<void> {
    await ensureLogOutNavElementExists();
    await playwrightHelperInstance!.waitForElementToAppear(REPORTS_NAV_ELEMENT);
    await playwrightHelperInstance!.findElementAndPerformAction(REPORTS_NAV_ELEMENT, 'click');
    await playwrightHelperInstance!.waitForElementToDisappear(PAGE_LOADING_ELEMENT);
    await playwrightHelperInstance!.waitForElementToAppear({ type: 'role', value: 'button', name: 'Create report' });
}

export async function checkVaccinesNavLinkExists(): Promise<boolean> {
    await ensureLogOutNavElementExists();
    await playwrightHelperInstance!.waitForElementToAppear(FIND_A_PATIENT_NAV_ELEMENT);
    return await playwrightHelperInstance!.checkElementExists(VACCINES_NAV_ELEMENT);
}

export async function clickVaccinesNavLink(): Promise<void> {
    await ensureLogOutNavElementExists();
    await playwrightHelperInstance!.waitForElementToAppear(VACCINES_NAV_ELEMENT);
    await playwrightHelperInstance!.findElementAndPerformAction(VACCINES_NAV_ELEMENT, 'click');
    await playwrightHelperInstance!.waitForElementToDisappear(PAGE_LOADING_ELEMENT);
    await playwrightHelperInstance!.waitForElementToAppear({ type: 'role', value: 'button', name: 'Add vaccine' });
}

export async function checkManageUsersNavLinkExists(): Promise<boolean> {
    await ensureLogOutNavElementExists();
    await playwrightHelperInstance!.waitForElementToAppear(FIND_A_PATIENT_NAV_ELEMENT);
    return await playwrightHelperInstance!.checkElementExists(MANAGE_USERS_NAV_ELEMENT);
}

export async function clickManageUsersNavLink(): Promise<void> {
    await ensureLogOutNavElementExists();
    await playwrightHelperInstance!.waitForElementToAppear(MANAGE_USERS_NAV_ELEMENT);
    await playwrightHelperInstance!.findElementAndPerformAction(MANAGE_USERS_NAV_ELEMENT, 'click');
    await playwrightHelperInstance!.waitForElementToDisappear(PAGE_LOADING_ELEMENT);
    await playwrightHelperInstance!.waitForElementToAppear({ type: 'role', value: 'button', name: 'Add user' });
}

export async function clickRecordVaccinationsNavLink(): Promise<void> {
  await ensureLogOutNavElementExists();
  await playwrightHelperInstance!.waitForElementToAppear(RECORD_VACCINATIONS_NAV_ELEMENT);
  await playwrightHelperInstance!.findElementAndPerformAction(RECORD_VACCINATIONS_NAV_ELEMENT, 'click');
  await playwrightHelperInstance!.waitForElementToDisappear(PAGE_LOADING_ELEMENT);
  await playwrightHelperInstance!.waitForElementToAppear({ type: 'role', value: 'button', name: 'Record vaccination' });
}

export async function checkNhsLogoNavElementExists(): Promise<boolean> {
  await ensureLogOutNavElementExists();
  return await playwrightHelperInstance!.checkElementExists(NHS_LOGO_NAV_ELEMENT);
}

export async function clickNhsLogoNavElement(): Promise<void> {
  await ensureLogOutNavElementExists();
  await playwrightHelperInstance!.findElementAndPerformAction(NHS_LOGO_NAV_ELEMENT, 'click');
  await playwrightHelperInstance!.waitForElementToDisappear(PAGE_LOADING_ELEMENT);
}

export async function getTotalVaccinationsHeadingText(): Promise<string> {
  await ensureLogOutNavElementExists();
  const text = await playwrightHelperInstance!.findElementAndPerformAction(TOTAL_VACCINATIONS_TEXT_ELEMENT, 'get_text');
  return text || '';
}

export async function getTodayVaccinationCountText(): Promise<string> {
  await ensureLogOutNavElementExists();
  const text = await playwrightHelperInstance!.findElementAndPerformAction(TODAY_VACCINATION_COUNT, 'get_text');
  return text || '';
}

export async function getWeekVaccinationCountText(): Promise<string> {
  await ensureLogOutNavElementExists();
  const text = await playwrightHelperInstance!.findElementAndPerformAction(WEEK_VACCINATION_COUNT, 'get_text');
  return text || '';
}

export async function getMonthVaccinationCountText(): Promise<string> {
  await ensureLogOutNavElementExists();
  const text = await playwrightHelperInstance!.findElementAndPerformAction(MONTH_VACCINATION_COUNT, 'get_text');
  return text || '';
}

export async function navigateToFeedbackPage(): Promise<void> {
  await ensureLogOutNavElementExists();
  await playwrightHelperInstance!.findElementAndPerformAction(FEEDBACK_LINK, 'click');
  await playwrightHelperInstance!.waitForElementToDisappear(PAGE_LOADING_ELEMENT);
}

export async function navigateToContactUsPage(): Promise<void> {
  await ensureLogOutNavElementExists();
  await playwrightHelperInstance!.findElementAndPerformAction(CONTACT_US_LINK, 'click');
  await playwrightHelperInstance!.waitForElementToDisappear(PAGE_LOADING_ELEMENT);
}

export async function navigateToHelpAndGuidancePage(): Promise<void> {
  await ensureLogOutNavElementExists();
  await playwrightHelperInstance!.findElementAndPerformAction(HELP_AND_GUIDANCE_LINK, 'click');
  await playwrightHelperInstance!.waitForElementToDisappear(PAGE_LOADING_ELEMENT);
}

export async function navigateToAccessibilityStatementPage(): Promise<void> {
  await ensureLogOutNavElementExists();
  await playwrightHelperInstance!.findElementAndPerformAction(ACCESSIBILITY_STATEMENT_LINK, 'click');
  await playwrightHelperInstance!.waitForElementToDisappear(PAGE_LOADING_ELEMENT);
}

export async function navigateToUserGuidePage(): Promise<void> {
  await ensureLogOutNavElementExists();
  await playwrightHelperInstance!.findElementAndPerformAction(USER_GUIDE_LINK, 'click');
  await playwrightHelperInstance!.waitForElementToDisappear(PAGE_LOADING_ELEMENT);
}

export async function navigateToTermsOfUsePage(): Promise<void> {
  await ensureLogOutNavElementExists();
  await playwrightHelperInstance!.findElementAndPerformAction(TERMS_OF_USE_LINK, 'click');
  await playwrightHelperInstance!.waitForElementToDisappear(PAGE_LOADING_ELEMENT);
}

export async function logoutFromApp(): Promise<void> {
  await ensureLogOutNavElementExists();
  await playwrightHelperInstance!.findElementAndPerformAction(LOGOUT_NAV_ELEMENT, 'click');
  await playwrightHelperInstance!.waitForElementToDisappear(PAGE_LOADING_ELEMENT);
}

export async function toggleNavbar(): Promise<void> {
  await ensureLogOutNavElementExists();
  await playwrightHelperInstance!.findElementAndPerformAction(NAV_BAR_TOGGLER, 'click');
}

export async function toggleNavLinkBar(): Promise<void> {
  await ensureLogOutNavElementExists();
  await playwrightHelperInstance!.findElementAndPerformAction(NAV_LINK_BAR_TOGGLER, 'click');
}

export async function checkLogoutButtonExistsOnProfilePage(): Promise<boolean> {
  await ensureLogOutNavElementExists();
  return await playwrightHelperInstance!.checkElementExists(LOGOUT_NAV_ELEMENT);
}

